tokenizers/src/models/unigram/trie.rs

```diff
@@ -30,23 +30,50 @@ impl<Label: Eq + Hash + Copy> Trie<Label> {
         node.is_leaf = true;
     }
 
-    pub fn common_prefix_search(&self, iterator: impl Iterator<Item = Label>) -> Vec<Vec<Label>> {
-        let mut node = &self.root;
-        let mut results = vec![];
-        let mut prefix = vec![];
-        for label in iterator {
-            prefix.push(label);
-            let child_opt = node.children.get(&label);
-            if let Some(child) = child_opt {
-                node = child;
-                if node.is_leaf {
-                    results.push(prefix.clone());
+    pub fn common_prefix_search<T>(&self, iterator: T) -> TrieIterator<Label, T>
+    where
+        T: Iterator<Item = Label>,
+    {
+        TrieIterator {
+            node: &self.root,
+            prefix: vec![],
+            iterator,
+        }
+    }
+}
+
+pub struct TrieIterator<'a, Label, T> {
+    node: &'a Node<Label>,
+    prefix: Vec<Label>,
+    iterator: T,
+}
+
+impl<Label, T> Iterator for TrieIterator<'_, Label, T>
+where
+    Label: Eq + Hash + Copy,
+    T: Iterator<Item = Label>,
+{
+    type Item = Vec<Label>;
```

Comment:
Well I had a hard time doing it, and the solution I found seemed unusable : https://stackoverflow.com/questions/30422177/how-do-i-write-an-iterator-that-returns-references-to-itself It does not work because now the vec is not owned anymore by the iterator, so I can't push into it anymore, that would require mutable ref, so I can't hand non mutable ref anymore.

Reply:
Alright, we can't indeed return a reference on the content of the Iterator itself.